### ISSUE TITLE
Panic in the `merge` method in the Allocator

### ISSUE THREAD

**Issue by Moxinilian** [slide-rs/specs#514]
I am not quite sure what is causing this panic as it is not my project, maybe @udoprog might know more.
This is in the context of https://github.com/udoprog/asteroids-amethyst/tree/44b7412591dd4e9c81cc771707c44debc63fad5c.
If you want to run it again, I recommend replacing amethyst in the Cargo.toml to:
```
amethyst = { git = "https://github.com/amethyst/amethyst.git", rev = "6dab277cba2837791ec77e8fec707618f3fb0d2e" }
```

**In debug mode**, if you spam the shooting button (space bar) for a long time and keep restarting (R button), eventually one of the restarts will trigger the following panic:
```
thread 'main' panicked at 'assertion failed: self.is_alive()', C:\Users\theod\.cargo\registry\src\github.com-1ecc6299db9ec823\specs-0.14.0\src\world\entity.rs:439:9
stack backtrace:
   0: std::sys::windows::backtrace::set_frames
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\sys\windows\backtrace\mod.rs:104
   1: std::sys::windows::backtrace::set_frames
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\sys\windows\backtrace\mod.rs:104
   2: std::sys_common::backtrace::_print
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\sys_common\backtrace.rs:71
   3: std::sys_common::backtrace::_print
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\sys_common\backtrace.rs:71
   4: std::panicking::default_hook::{{closure}}
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\panicking.rs:211
   5: std::panicking::default_hook
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\panicking.rs:227
   6: std::panicking::rust_panic_with_hook
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\panicking.rs:476
   7: std::panicking::begin_panic<str*>
             at \libstd\panicking.rs:410
   8: specs::world::entity::ZeroableGeneration::die
             at C:\Users\theod\.cargo\registry\src\github.com-1ecc6299db9ec823\specs-0.14.0\src\world\entity.rs:439
   9: specs::world::entity::Allocator::merge
             at C:\Users\theod\.cargo\registry\src\github.com-1ecc6299db9ec823\specs-0.14.0\src\world\entity.rs:186
  10: specs::world::World::maintain
             at C:\Users\theod\.cargo\registry\src\github.com-1ecc6299db9ec823\specs-0.14.0\src\world\mod.rs:560
  11: amethyst::app::CoreApplication<asteroids_amethyst::states::Data, amethyst::state_event::StateEvent, amethyst::state_event::StateEventReader>::advance_frame<asteroids_amethyst::states::Data,amethyst::state_event::StateEvent,amethyst::state_event::StateEventReader>
             at C:\Users\theod\.cargo\git\checkouts\amethyst-fedb0a1032a075ce\6dab277\src\app.rs:370
  12: amethyst::app::CoreApplication<asteroids_amethyst::states::Data, amethyst::state_event::StateEvent, amethyst::state_event::StateEventReader>::run<asteroids_amethyst::states::Data,amethyst::state_event::StateEvent,amethyst::state_event::StateEventReader>
             at C:\Users\theod\.cargo\git\checkouts\amethyst-fedb0a1032a075ce\6dab277\src\app.rs:232
  13: asteroids_amethyst::main
             at .\src\main.rs:100
  14: std::rt::lang_start::{{closure}}<core::result::Result<(), amethyst::error::Error>>
             at \libstd\rt.rs:74
  15: std::rt::lang_start_internal::{{closure}}
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\rt.rs:59
  16: std::rt::lang_start_internal::{{closure}}
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\rt.rs:59
  17: panic_unwind::__rust_maybe_catch_panic
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libpanic_unwind\lib.rs:102
  18: std::panicking::try
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\panicking.rs:289
  19: std::panicking::try
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\panicking.rs:289
  20: std::panicking::try
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\panicking.rs:289
  21: std::rt::lang_start<core::result::Result<(), amethyst::error::Error>>
             at \libstd\rt.rs:74
  22: main
  23: invoke_main
             at f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl:78
  24: invoke_main
             at f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl:78
  25: BaseThreadInitThunk
26: RtlUserThreadStart
```

**In release mode**, with the same steps, you get a different panic that probably is related (considering the previous panic was a debug_assert):
```
thread 'main' panicked at 'assertion failed: !self.is_alive()', C:\Users\theod\.cargo\registry\src\github.com-1ecc6299db9ec823\specs-0.14.0\src\world\entity.rs:449:9
stack backtrace:
   0: std::sys::windows::backtrace::set_frames
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\sys\windows\backtrace\mod.rs:104
   1: std::sys::windows::backtrace::set_frames
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\sys\windows\backtrace\mod.rs:104
   2: std::sys_common::backtrace::_print
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\sys_common\backtrace.rs:71
   3: std::sys_common::backtrace::_print
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\sys_common\backtrace.rs:71
   4: std::panicking::default_hook::{{closure}}
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\panicking.rs:211
   5: std::panicking::default_hook
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\panicking.rs:227
   6: std::panicking::rust_panic_with_hook
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\panicking.rs:476
   7: std::panicking::begin_panic
   8: specs::world::entity::Allocator::merge
   9: specs::world::World::maintain
  10: <amethyst::app::CoreApplication<'a, T, E, R>>::run
  11: asteroids_amethyst::textures::SpriteSheet::from_path
  12: ron::parse::Bytes::float
  13: std::rt::lang_start_internal::{{closure}}
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\rt.rs:59
  14: std::rt::lang_start_internal::{{closure}}
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\rt.rs:59
  15: panic_unwind::__rust_maybe_catch_panic
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libpanic_unwind\lib.rs:102
  16: std::panicking::try
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\panicking.rs:289
  17: std::panicking::try
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\panicking.rs:289
  18: std::panicking::try
             at /rustc/424a749a01224239ba2c8850f16007d57db0a242\src/libstd\panicking.rs:289
  19: main
  20: invoke_main
             at f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl:78
  21: invoke_main
             at f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl:78
  22: BaseThreadInitThunk
23: RtlUserThreadStart
```
I believe the panic is from the assert at [`src/world/entity.rs#L410`](https://github.com/slide-rs/specs/blob/f0b7377c8f028b62545d76095fd31009393ea6fb/src/world/entity.rs#L410), but not explicit in the backtrace as they probably get inlined.

I am not entirely sure why those asserts are present in allocator however. If they only are about bugs like that, maybe they should either be more explicit or replaced by more typical bug-related panics.

Note that the panic also happens when you are not shooting, but takes many more restarts before happening. If I did not know better, I could say it is a generation overflow, but it seems unreasonable here. I'd suggest looking at how the dead entity stack in the allocator evolves to see if something fishy happens there.

**Comment by udoprog** [comment]
Can verify that I'm seeing it on Linux now as well.

**Comment by udoprog** [comment]
One thing to note is that I make liberal use of `LazyUpdate`, and in particular [`world.maintain()`](https://github.com/amethyst/amethyst/blob/master/src/app.rs#L370) might not be called before `world.delete_all()` since I'm [dispatching right before checking for a restart](https://github.com/udoprog/asteroids-amethyst/blob/master/src/states.rs#L101).

I was not able to create a reproducible example out of this though :/. I might add some code to the game to play itself for a while so that it's easier to reproduce.

**Comment by udoprog** [comment]
Note: all the ordering in the atomic bitsets seems to be [`Relaxed`](https://github.com/slide-rs/hibitset/blob/master/src/atomic.rs#L156), which includes when reading the layers for iteration.

Ordering is really hard to reason about though, but IIUC: Without more constraints, the insertions and iterations _might_ be re-ordered at compile time. Especially since [all bitset operations are inlined](https://github.com/slide-rs/hibitset/blob/master/src/atomic.rs#L82) and if the compiler is clever enough to realize that e.g. `remove` doesn't actually require mutable access.

I'm far from certain about the above, but it could be debunked by looking at the code being generated. Replacing all Ordering with `SeqCst` might also be beneficial.

**Comment by torkleyy** [comment]
Investigating

**Comment by torkleyy** [comment]
Successfully reproduced this. Initial thoughts:

* restarting logic is the origin of the issue
* in particular, trying to shoot while we're restarting triggers warnings (`LazyUpdate`) and panics

**Comment by torkleyy** [comment]
Fixed the bug in #533.

@udoprog As for the warnings, that's a side effect of using `LazyUpdate`.
There are two solutions:

* Ignore failed deletion silently
* Maintain the world before deleting all entities
